usb/device/USBCDC_ECM/USBCDC_ECM.cpp

```diff
@@ -0,0 +1,496 @@
+        ENDPOINT_DESCRIPTOR,            // bDescriptorType
+        _int_in,                        // bEndpointAddress
+        E_INTERRUPT,                    // bmAttributes (0x03=intr)
+        (uint8_t) LSB(MAX_PACKET_SIZE_INT),     // wMaxPacketSize (LSB)
+        (uint8_t) MSB(MAX_PACKET_SIZE_INT),     // wMaxPacketSize (MSB)
+        16,                             // bInterval
+
+        // Default data interface descriptor, USB spec 9.6.5, page 267-269, Table 9-12
+        0x09,                       // bLength
+        INTERFACE_DESCRIPTOR,       // bDescriptorType
+        1,                          // bInterfaceNumber
+        0,                          // bAlternateSetting
+        0,                          // bNumEndpoints
+        0x0A,                       // bInterfaceClass
+        0x00,                       // bInterfaceSubClass
+        0x00,                       // bInterfaceProtocol
+        0,                          // iInterface
+
+        // Data interface descriptor, USB spec 9.6.5, page 267-269, Table 9-12
+        0x09,                       // bLength
+        INTERFACE_DESCRIPTOR,       // bDescriptorType
+        1,                          // bInterfaceNumber
+        1,                          // bAlternateSetting
+        2,                          // bNumEndpoints
+        0x0A,                       // bInterfaceClass
+        0x00,                       // bInterfaceSubClass
+        0x00,                       // bInterfaceProtocol
+        0,                          // iInterface
+
+        // Endpoint descriptor, USB spec 9.6.6, page 269-271, Table 9-13
+        ENDPOINT_DESCRIPTOR_LENGTH, // bLength
+        ENDPOINT_DESCRIPTOR,        // bDescriptorType
+        _bulk_in,                   // bEndpointAddress
+        E_BULK,                     // bmAttributes (0x02=bulk)
+        (uint8_t) LSB(MAX_PACKET_SIZE_BULK),  // wMaxPacketSize (LSB)
+        (uint8_t) MSB(MAX_PACKET_SIZE_BULK),  // wMaxPacketSize (MSB)
+        0,                          // bInterval
+
+        // Endpoint descriptor, USB spec 9.6.6, page 269-271, Table 9-13
+        ENDPOINT_DESCRIPTOR_LENGTH, // bLength
+        ENDPOINT_DESCRIPTOR,        // bDescriptorType
+        _bulk_out,                  // bEndpointAddress
+        E_BULK,                     // bmAttributes (0x02=bulk)
+        (uint8_t) LSB(MAX_PACKET_SIZE_BULK),  // wMaxPacketSize (LSB)
+        (uint8_t) MSB(MAX_PACKET_SIZE_BULK),  // wMaxPacketSize (MSB)
+        0                           // bInterval
+    };
+    return configDescriptor;
+}
+
+void USBCDC_ECM::_int_callback()
+{
+    assert_locked();
+
+    _flags.set(FLAG_INT_DONE);
+}
+
+void USBCDC_ECM::_bulk_in_callback()
+{
+    assert_locked();
+
+    _flags.set(FLAG_WRITE_DONE);
+}
+
+void USBCDC_ECM::_bulk_out_callback()
+{
+    assert_locked();
+
+    _bulk_buf_size = read_finish(_bulk_out);
+
+    read_start(_bulk_out, _bulk_buf, MAX_PACKET_SIZE_BULK);
```

Comment:
Without being able to receive data the functionality of this device is quite limited.